File: src/main/java/net/eduard/api/lib/bungee/BungeeAPI.java
package net.eduard.api.lib.bungee;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class BungeeAPI {
    BungeeAPI() {
    }

    private final static String channel = "bukkit:bungee";
    private final static Map<String, ServerSpigot> servers = new HashMap<>();
    private final static List<ServerMessageHandler> handlers = new ArrayList<>();
    private static ServerController controller;


    public static ServerSpigot getServer(String serverName) {
        ServerSpigot server = servers.get(serverName.toLowerCase());
        if (server == null) {
            server = new ServerSpigot(serverName);
            servers.put(serverName.toLowerCase(), server);
        }
        return server;
    }

    public static int getPlayersAmount() {
        int amount = 0;
        for (ServerSpigot server : servers.values()) {
            amount+=server.getCount();
        }
        return amount;
    }

    public static int getPlayersAmount(String serverType) {
        int amount = 0;
        for (ServerSpigot server : servers.values()) {
            if (server.getType().equalsIgnoreCase(serverType)) {
                amount += server.getCount();
            }
        }
        return amount;
    }

    public static int getPlayersAmount(String serverType, int teamSize) {
        int amount = 0;
        for (ServerSpigot server : servers.values()) {
            if (!server.getSubType().equalsIgnoreCase("lobby")
                    && server.getType().equalsIgnoreCase(serverType)
                    && server.getTeamSize() == teamSize) {
                amount += server.getCount();
            }
        }
        return amount;
    }

    public static int getPlayersAmount(String serverType, String subType) {
        int amount = 0;
        for (ServerSpigot server : servers.values()) {
            if (server.getType().equalsIgnoreCase(serverType) &&
                    subType.equalsIgnoreCase(server.getSubType())) {
                amount += server.getCount();
            }
        }
        return amount;
    }

    static {
        try {
            Class.forName("org.bukkit.Bukkit");
            controller = new BukkitController();
        } catch (Exception e) {
            controller = new BungeeController();
        }
    }

    public static BukkitController getBukkit() {
        return (BukkitController) controller;
    }

    public static BungeeController getBungee() {
        return (BungeeController) controller;
    }

    public static String getChannel() {
        return channel;
    }

    public static void register(ServerMessageHandler receiver) {
        handlers.add(receiver);
    }

    public static List<ServerMessageHandler> getHandlers() {
        return handlers;
    }

    public static ServerController getController() {
        return controller;
    }

    public static Map<String, ServerSpigot> getServers() {
        return servers;
    }


}

File: src/main/java/net/eduard/api/lib/bungee/BungeeAPI.kt
package net.eduard.api.lib.bungee

import net.eduard.api.lib.hybrid.Hybrid
import java.lang.Exception

object BungeeAPI {

    fun debug(msg : String){

        Hybrid.instance.console.sendMessage("§b[BungeeAPI] §f$msg")

    }

    const val channel = "bukkit:bungee"
    val servers = mutableMapOf<String, ServerSpigot>()
    val handlers = mutableListOf<ServerMessageHandler>()


    fun getServer(serverName: String): ServerSpigot {
        return servers.getOrPut(serverName.toLowerCase() , {
            ServerSpigot(serverName)
        })
    }

    val playersAmount: Int
        get() {
            var amount = 0
            for (server in servers.values) {
                amount += server.count
            }
            return amount
        }

    fun getPlayersAmount(serverType: String): Int {
        var amount = 0
        for (server in servers.values) {
            if (server.type.equals(serverType, ignoreCase = true)) {
                amount += server.count
            }
        }
        return amount
    }

    fun getPlayersAmount(serverType: String, teamSize: Int): Int {

        var amount = 0
        for (server in servers.values) {
            if (!server.subType.equals("lobby", ignoreCase = true)
                && server.type.equals(serverType, ignoreCase = true)
                && server.teamSize == teamSize
            ) {
                amount += server.count
            }
        }

        return amount
    }

    fun getPlayersAmount(serverType: String, subType: String): Int {
        var amount = 0
        for (server in servers.values) {
            if (server.type.equals(serverType, ignoreCase = true) &&
                subType.equals(server.subType, ignoreCase = true)
            ) {
                amount += server.count
            }
        }
        return amount
    }

    val bukkit: BukkitController
        get() = controller as BukkitController
    val bungee: BungeeController
        get() = controller as BungeeController

    fun register(receiver: ServerMessageHandler) {
        handlers.add(receiver)
    }
    var controller: ServerController<*> = try {
        Class.forName("org.bukkit.Bukkit")
        BukkitController()
    } catch (e: Exception) {
        BungeeController()
    }

}